src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -276,6 +276,24 @@ public void Remove(object key)
             StartScanForExpiredItemsIfNeeded(_options.Clock.UtcNow);
         }
 
+        /// <summary>
+        /// Removes all keys and values from the cache.
+        /// </summary>
+        public void Clear()
+        {
+            CheckDisposed();
+
+            // the following two operations are not atomic change as a whole, but an alternative would be to introduce a global lock for every access to _entries and _cacheSize
```

Comment:
Could we get away with this by introducing a new class that contained both the dictionary and the size? Then we can atomically replace the pointer with a new instance?

Reply:
In this particular case, yes. In other, where we add/remove item from the cache and update the size afterwards, not.

https://github.com/dotnet/runtime/blob/40cbe82e194d4f0574c9796f1c5aff4c17eac55b/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L265-L270

https://github.com/dotnet/runtime/blob/40cbe82e194d4f0574c9796f1c5aff4c17eac55b/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L151-L158

I am not sure if it's worth it.

Reply:
In those other cases you can atomically read the pointer, store it in a local variable and modify that. If Clear gets called concurrently, it will either update the pointer before or after the other mutation. Thus, we should always have a coherent set of entries and cache size.

Reply:
@eerhardt well, it took me a while to get back to this PR. PTAL at my recent commit.